builtin/logical/ad/path_config.go

```diff
@@ -0,0 +1,219 @@
+				Default:     "tls12",
+				Description: "Maximum TLS version to use. Accepted values are 'tls10', 'tls11' or 'tls12'. Defaults to 'tls12'",
+			},
+
+			"default_password_ttl": {
+				Type:        framework.TypeInt,
+				Default:     defaultPasswordTTLs,
+				Description: "In minutes, the default password time-to-live.",
+			},
+
+			"max_password_ttl": {
+				Type:        framework.TypeInt,
+				Default:     defaultPasswordTTLs,
+				Description: "In minutes, the maximum password time-to-live.",
+			},
+
+			"password_length": {
+				Type:        framework.TypeInt,
+				Default:     defaultPasswordLength,
+				Description: "The desired length of passwords that Vault generates.",
+			},
+		},
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.DeleteOperation: h.DeleteOperation,
+			logical.ReadOperation:   h.ReadOperation,
+			logical.UpdateOperation: h.UpdateOperation,
+		},
+	}
+}
+
+// Config is a convenience method for other operations to use in retrieving the current config.
+func (h *configHandler) Config(ctx context.Context, req *logical.Request) (*engineConfig, error) {
+
+	entry, err := req.Storage.Get(ctx, "config")
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return &engineConfig{}, nil
+	}
+
+	engineConf := &engineConfig{}
+	if err := entry.DecodeJSON(engineConf); err != nil {
+		return nil, err
+	}
+
+	return engineConf, nil
+}
+
+func (h *configHandler) DeleteOperation(ctx context.Context, req *logical.Request, _ *framework.FieldData) (*logical.Response, error) {
+	if err := req.Storage.Delete(ctx, "config"); err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+func (h *configHandler) ReadOperation(ctx context.Context, req *logical.Request, _ *framework.FieldData) (*logical.Response, error) {
+
+	engineConf, err := h.Config(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &logical.Response{
+		Data: engineConf.Map(),
+	}
+	resp.AddWarning("Read access to this endpoint should be controlled via ACLs as it will return the configuration information as-is, including any passwords.")
+	return resp, nil
+}
+
+func (h *configHandler) UpdateOperation(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+
+	activeDirectoryConf, err := activedirectory.NewConfiguration(h.logger, fieldData)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), err
+	}
+	passwordConf := newPasswordConfig(fieldData)
+	engineConf := &engineConfig{passwordConf, activeDirectoryConf}
+
+	entry, err := logical.StorageEntryJSON("config", engineConf)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := req.Storage.Put(ctx, entry); err != nil {
+		return nil, err
+	}
+
+	resp := &logical.Response{
+		Data: engineConf.Map(),
+	}
+	resp.AddWarning("Write access to this endpoint should be controlled via ACLs as it will return the configuration information as-is, including any passwords.")
+	return resp, nil
+}
+
+type engineConfig struct {
```

Comment:
I will add and use a `Validate` method to this `engineConf` that checks for required fields specific to this backend.

Reply:
Also, I'm going to add a constructor that fulfills the pointers and tracks whether it's been set, so that there's no way to get a nil pointer from this thing.